pkg/operator/resources/cluster/apiserver.go

```diff
@@ -611,6 +659,78 @@ func createDataVolumeMutatingWebhook(namespace string, c client.Client, l logr.L
 	return whc
 }
 
+func createPvcMutatingWebhook(namespace string, c client.Client, l logr.Logger) *admissionregistrationv1.MutatingWebhookConfiguration {
+	path := "/pvc-mutate"
+	defaultServicePort := int32(443)
+	allScopes := admissionregistrationv1.AllScopes
+	exactPolicy := admissionregistrationv1.Exact
+	failurePolicy := admissionregistrationv1.Ignore //unlike DV
```

Comment:
Here's the problem we've encountered with creating webhooks for core k8s apis (like pods, pvcs, nodes, etc...)

If the webhook isn't required, then that can lead to unpredictable results for users. Sometimes they get the thing they want, sometimes they don't 

If the webhook is required, then we've created a component that directly impacts the availability of the entire cluster, beyond kubevirt and cdi. That's not a place we want to be in.

This is why we've avoided mutating core k8s apis.

Reply:
> Here's the problem we've encountered with creating webhooks for core k8s apis (like pods, pvcs, nodes, etc...)
> 
> If the webhook isn't required, then that can lead to unpredictable results for users. Sometimes they get the thing they want, sometimes they don't
> 
> If the webhook is required, then we've created a component that directly impacts the availability of the entire cluster, beyond kubevirt and cdi. That's not a place we want to be in.
> 
> This is why we've avoided mutating core k8s apis.

The webhook intercepts only explicitly CDI-labeled PVCs  (`cdi.kubevirt.io/useStorageProfile: true`) and ignores others. The DV mutating webhook was always required, so I don't see setting the PVC mutating webhook required as a potential cluster availability regression. If webhook is not required, PVC creation may fail as it will miss required fields, and I think this is also not a big deal, as it wil be reconciled again.

Reply:
is it possible to make the webhook required, but only impact creation of pvcs with the special label when the webhook is not available?

Reply:
> is it possible to make the webhook required, but only impact creation of pvcs with the special label when the webhook is not available?

```
In the event of an error evaluating a match condition the webhook is never called. Whether to reject the request is determined as follows:

If any match condition evaluated to false (regardless of other errors), the API server skips the webhook.
Otherwise:
for failurePolicy: Fail, reject the request (without calling the webhook).
for failurePolicy: Ignore, proceed with the request but skip the webhook.
```
see: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/

Reply:
in that case, I'm less concerned about a mutating webhook on a core api impacting cluster availability. I would recommend using `failurePolicy: Fail`. consistent behavior is important.

All that said, I'm conflicted about mutating core APIs like this. This can lead to subtle issues for users. Someone creates a PVC with the magic label and things appear to work. Then someone uses the spec in that PVC somewhere else and it doesn't work because there's no magic label. Reducing "magic" is what well structured and predictable APIs give us. 

In general, I'd advise against mutating core APIs, but this is a nuanced topic that depends on the team's culture and use case.

Reply:
Note this is the piece that will give VolumeClaimTemplates equivalence to DataVolumeTemplates. If we decided that we want to continue with VolumeClaimTemplates. We would really like to get to the point where DataVolumes are not needed at all anymore, but we need some mechanism to provide some of the nice things that DataVolumes provide.

Also note my reaction was, mutating core k8s resources is a bad idea in general, but I think with the label restrictions it won't affect cluster stability if the CDI api server is down. As well as the fact this is a PVC, not a pod, which should get created a lot less than pods.